### PR TITLE
feat: support native social sign using apple sdk

### PR DESCRIPTION
**Pull request by jonas-jonas**
How this works:

1. Use the Apple SDK on the device to request authentication, which returns an IDToken
2. Submit that token to Kratos using updateRegistrationFlow, alongside any traits you want the identity to have
3. Kratos verifies the token using Apple's public JWKS (make this configurable for the generic provider?)
4. Kratos extracts the data from the IDToken
5. Kratos processes the login/registration as if it came from an OIDC callback

Expo example (_PR in repo following_):
```ts
import * as AppleAuthentication from "expo-apple-authentication"
import * as Crypto from "expo-crypto"
import * as Random from "expo-random"

async function signInWithApple() {
  const nonce = Random.getRandomBytes(16).toString()

  const digest = await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    nonce,
  )
  const credential = await AppleAuthentication.signInAsync({
    requestedScopes: [
      AppleAuthentication.AppleAuthenticationScope.EMAIL,
      AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
    ],
    nonce: digest,
  })

  return orySdk.updateRegistrationFlow({
    flow: flow.id,
    updateRegistrationFlowBody: {
      provider: "apple",
      id_token: credential.identityToken,
      raw_id_token_nonce: nonce,
      traits: {
        name: {
          first: credential.fullName?.givenName || "given name", // When developing, these values might be empty
          last: credential.fullName?.familyName || "last name", // When developing, these values might be empty
        },
      },
    },
  })
}
```
## Related issue(s)

Docs PR: https://github.com/ory/docs/pull/1528

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: selfservice/strategy/oidc/provider_apple.go
Comment (jonas-jonas) [inline selfservice/strategy/oidc/provider_apple.go:153]: One limitation of this implementation is that we don't request ATs & RTs during this flow, so if users want to use them, on behalf of their users, this wouldn't work. 

I am not sure what use cases there are for using an AT against Apple's API, but I'd suggest we can implement the `authCode` flow at a later time. It should be trivial, but [doesn't work with the way Expo has implemented Apple authentication](https://github.com/expo/expo/issues/16162), as it requests the tokens using the client ID `host.exp.Exponent`.

File: selfservice/strategy/oidc/provider_apple.go
Comment (jonas-jonas) [inline selfservice/strategy/oidc/provider_apple.go:156]: This is the only reason you'd need a configured apple client in the OIDC config, as it checks whether the IDToken was issued to the same `client_id`. As mentioned above, this means that for development using expo, you'd need to set it to `host.exp.Exponent`. 

I am not 100% sure this is secure enough, but since we verify that the token was issued by Apple, it _should_ be fine™️.

In production apps, the token would be issued to `sh.ory.iosapp` on the device, and the Ory project/configuration would need a client with the same `client_id`. `apple_team_id`, `apple_private_key_id` an `apple_private_key` would be ignored in this flow.

File: selfservice/strategy/oidc/provider_config.go
Comment (jonas-jonas) [inline selfservice/strategy/oidc/provider_config.go:124]: This refactoring is needed to be able to mock a new provider in that can selectively verify or fail verification of an ID Token.

I've double-checked, that I migrated all providers, but *please* check this again! Thank you.

File: selfservice/strategy/oidc/provider_apple.go
Comment (jonas-jonas) [inline selfservice/strategy/oidc/provider_apple.go:166]: Any ideas on how to test this are appreciated. I thought about adding a test that just uses a real `idToken` and verifies that, but that token contains my personal email address and user ID, which I'd rather not leak here. :)

I guess we could create a dummy Apple ID, obtain the token and add that here, however we would need to refresh the token, as Apple rotates their JWKs. Also, not sure how to deal with expiry.

**Review by jonas-jonas — COMMENTED** [review]
I am not sure, we can really e2e test this, as there is no other provider that supports verifying an ID Token from the payload, for now. 

I guess, we could make the generic provider support it and add a JWKS URL to the config? Then we could obtain an ID token from a Hydra instance and submit that.

File: selfservice/strategy/oidc/provider_apple.go
Comment (aeneasr) [reply selfservice/strategy/oidc/provider_apple.go:166]: Generate a JWT with the same claims as apple, and set up a fake HTTP server with a JWK you control - then verify. That should work!

File: selfservice/strategy/oidc/provider_apple.go
Comment (aeneasr) [reply selfservice/strategy/oidc/provider_apple.go:153]: Please document this as a caveat.

**Review by aeneasr — COMMENTED** [review]
Looks very good, please add three tests to strategy_test:

- [ ] Test login with an ID token, ensure claims end up properly
- [ ] Test registration with an ID token, ensure claims end up properly
- [ ] Test login/registration with an ID token where the Client ID is not matching (should fail)
- [ ] Test login/registration with an ID token where the nonce is not matching (should fail)

Please also make it possible to submit the nonce and validate the nonce as well. Fail the flow if the nonce is not set.

File: selfservice/strategy/oidc/provider_apple.go
Comment (jonas-jonas) [reply selfservice/strategy/oidc/provider_apple.go:169]: Yup, that works. I needed to make the JWKS URL configurable via a private field, which defaults to the apple URL.

File: selfservice/strategy/oidc/provider_apple.go
Comment (jonas-jonas) [reply selfservice/strategy/oidc/provider_apple.go:156]: Done in the Docs PR.

File: selfservice/strategy/oidc/provider.go
Comment (aeneasr) [inline selfservice/strategy/oidc/provider.go:61]: Is this an actual claim returned by Apple/Google?

File: selfservice/strategy/oidc/strategy.go
Comment (aeneasr) [inline selfservice/strategy/oidc/strategy.go:625]: Why is this being hex encoded? From the OIDC specification:

> **nonce:** String value used to associate a Client session with an ID Token, and to mitigate replay attacks. The value is passed through unmodified from the Authentication Request to the ID Token. If present in the ID Token, Clients MUST verify that the nonce Claim Value is equal to the value of the nonce parameter sent in the Authentication Request. If present in the Authentication Request, Authorization Servers MUST include a nonce Claim in the ID Token with the Claim Value being the nonce value sent in the Authentication Request. Authorization **Servers SHOULD perform no other processing on nonce values used. The nonce value is a case sensitive string.**

https://openid.net/specs/openid-connect-core-1_0.html

https://github.com/coreos/go-oidc/blob/f0117a5f178cb36710dde42ede2154071b20ee94/example/idtoken/app.go#L109-L117

If this check is Apple-specific, it should live in the apple strategy IMO.

File: selfservice/strategy/oidc/strategy_login.go
Comment (aeneasr) [inline selfservice/strategy/oidc/strategy_login.go:101]: Why is this prefixed `raw_`? IMO it should be `id_token_nonce`

File: selfservice/strategy/oidc/strategy_test.go
Comment (aeneasr) [inline selfservice/strategy/oidc/strategy_test.go:632]: Missing test: nonce mismatch

File: selfservice/strategy/oidc/provider_apple_test.go
Comment (aeneasr) [inline selfservice/strategy/oidc/provider_apple_test.go:18]: Please use `golang-jwt/jwt-go/v4`

File: selfservice/strategy/oidc/provider.go
Comment (aeneasr) [reply selfservice/strategy/oidc/provider.go:61]: I see, this is returned by Apple. I don't think this is a standard claim but rather something specific to Apple. In that case, I would suggest to only handle this claim in the apple provider. The problem here is that this bool value will default to "false", which disables the nonce check if the value is not available (which is the case with any other platform). So basically, this is a bypass for the nonce check, built in and enabled per default.

**Review by aeneasr — CHANGES_REQUESTED** [review]
The nonce implementation does not look correct to me. Typically, a nonce is a plain string (like the state parameter) and you do a string equality match to verify if it is valid or not. The `nonce_supported` flag defaults to false, which *may* be OK for apple, but is definitely not OK for any other provider, as this claim is not a standard claim (it is never mentioned here: https://openid.net/specs/openid-connect-core-1_0.html).

Additionally, the Apple documentation states:

> A Boolean value that indicates whether the transaction is on a nonce-supported platform. If you send a nonce in the authorization request, but don’t see the nonce claim in the identity token, check this claim to determine how to proceed. If this claim returns true, treat nonce as mandatory and fail the transaction; otherwise, you can proceed treating the nonce as optional.

However, as far as I can see, we use this claim first to check whether we should validate. But the documentation states:

1. If a nonce was sent (aka if we receive the nonce as part of the registration / login payload
2. And the nonce is not available in the id_token
3. Then (and only then) we check this claim

The current implementation however:

1. Checks if the claim is true
2. Proceeds with validation

My suggestion is to move this custom validation logic into the apple ID token verification:

1. The apple provider validates the `nonce` claim:
  1. If a nonce was sent as part of the payload
  2. And the id_token has a nonce -> we validate
  3. And the id_token has no nonce
    * We check the nonce_supported claim
    * If true: fail
    * If false: proceed
2. The Google provider also validates the nonce (but always, because Google returns this always)
3. The generic provider does not validate the nonce at all, it delegates the ID token validation to the specific provider (we can fix this later if desired).

Please don't merge this without another review from myself!

File: selfservice/strategy/oidc/strategy.go
Comment (jonas-jonas) [reply selfservice/strategy/oidc/strategy.go:625]: Fixed. Thanks for the sources.

File: selfservice/strategy/oidc/strategy_login.go
Comment (jonas-jonas) [reply selfservice/strategy/oidc/strategy_login.go:101]: Fixed, misunderstand of how nonces work on my part.

File: selfservice/strategy/oidc/provider_apple_test.go
Comment (jonas-jonas) [reply selfservice/strategy/oidc/provider_apple_test.go:18]: Done.

File: selfservice/strategy/oidc/strategy_test.go
Comment (jonas-jonas) [reply selfservice/strategy/oidc/strategy_test.go:632]: Fixed, nice catch.

File: selfservice/strategy/oidc/strategy.go
Comment (jonas-jonas) [inline selfservice/strategy/oidc/strategy.go:610]: We could also just require nonces to be present in the token regardless of the provider. If someone wants to integrate with an Apple device that does not support the nonce, we can still add this in later.

**Review by aeneasr — COMMENTED** [review]
Much better! Almost there

File: selfservice/strategy/oidc/strategy.go
Comment (jonas-jonas) [reply selfservice/strategy/oidc/strategy.go]: Done! The check is a bit less readable now, if you have an idea on improvement, LMK.

**Review by aeneasr — APPROVED** [review]
Awesome!

**Review by hperl — APPROVED** [review]

**Pull request merged**
